formats/cbor/commonMain/src/kotlinx/serialization/cbor/internal/Encoder.kt

```diff
@@ -177,10 +182,207 @@ internal class IndefiniteLengthCborWriter(cbor: Cbor, output: ByteArrayOutput) :
     override fun incrementChildren() {/*NOOP*/
     }
 
+    override fun encodeTags(tags: ULongArray) = tags.forEach { getDestination().encodeTag(it) }
+
+}
+
+// optimized indefinite length encoder
+internal class StructuredCborWriter(cbor: Cbor) : CborWriter(cbor) {
+
+    /**
+     * Tags and values are "written", i.e. recorded/prepared for encoding separately. Hence, we need a helper that allows
+     * for setting tags and values independently, and then merging them into the final [CborElement] at the end.
+     */
+    internal sealed class CborContainer(tags: ULongArray, elements: MutableList<CborElement>) {
+        protected val elements = elements
+
+        var tags = tags
+            internal set
+
+
+        open fun add(element: CborElement) = elements.add(element)
+        class Map(tags: ULongArray, elements: MutableList<CborElement> = mutableListOf()) :
+            CborContainer(tags, elements)
+
+        class List(tags: ULongArray, elements: MutableList<CborElement> = mutableListOf()) :
+            CborContainer(tags, elements)
+
+        class Primitive(tags: ULongArray) : CborContainer(tags, elements = mutableListOf()) {
+            override fun add(element: CborElement): Boolean {
+                require(elements.isEmpty()) { "Implementation error. Please report a bug." }
+                return elements.add(element)
+            }
+        }
+
+        fun finalize() = when (this) {
+            is List -> CborList(content = elements, tags = tags)
+            is Map -> CborMap(
+                content = if (elements.isNotEmpty()) IntRange(0, elements.size / 2 - 1).associate {
+                    elements[it * 2] to elements[it * 2 + 1]
+                } else mapOf(),
```

Comment:
I would also vote for 
```kotlin
buildMap {
    for (i in 0 until elements.size / 2 - 1) {
        put(elements[i * 2], elements[i * 2 + 1])
     }
}
```
as an alternative to `if (..) associate {} or mapOf()`